source/extensions/filters/common/expr/custom_cel/extended_request/extended_request_cel_vocabulary.cc

```diff
@@ -0,0 +1,166 @@
+                   descriptor.name(), status.message());
+  }
+}
+
+void addStaticFunctionToRegistry(CelFunctionRegistry* registry,
+                                 std::unique_ptr<CelFunction> function) {
+  auto function_name = function->descriptor().name();
+  absl::Status status = registry->Register(std::move(function));
+  if (!status.ok()) {
+    ENVOY_LOG_MISC(debug, "Failed to register static function {}  in CEL function registry: {}",
+                   function_name, status.message());
+  }
+}
+
+void ExtendedRequestCelVocabulary::fillActivation(
+    Activation* activation, Protobuf::Arena& arena, const StreamInfo::StreamInfo& info,
+    const Http::RequestHeaderMap* request_headers, const Http::ResponseHeaderMap* response_headers,
+    const Http::ResponseTrailerMap* response_trailers) {
+  request_headers_ = request_headers;
+  response_headers_ = response_headers;
+  response_trailers_ = response_trailers;
+
+  // all of the following require request_headers
+  if (request_headers) {
+    // attributes
+    addValueProducerToActivation(
+        activation, ExtendedRequest,
+        std::make_unique<ExtendedRequestWrapper>(arena, request_headers, info,
+                                                 return_url_query_string_as_map_));
+
+    // lazy functions only
+    addLazyFunctionToActivation(
+        activation, LazyFuncNameCookie, false,
+        std::function<CelValue(Protobuf::Arena*)>(
+            [request_headers](Protobuf::Arena* arena) { return cookie(arena, *request_headers); }));
+    addLazyFunctionToActivation(activation, LazyFuncNameCookieValue, false,
+                                std::function<CelValue(Protobuf::Arena*, CelValue)>(
+                                    [request_headers](Protobuf::Arena* arena, CelValue key) {
+                                      return cookieValue(arena, *request_headers, key);
+                                    }));
+  }
+}
+
+void ExtendedRequestCelVocabulary::registerFunctions(CelFunctionRegistry* registry) {
+  absl::Status status;
+
+  // lazy functions
+  addLazyFunctionToRegistry(registry, CelFunctionDescriptor(LazyFuncNameCookie, false, {}));
+  addLazyFunctionToRegistry(
+      registry, CelFunctionDescriptor(LazyFuncNameCookieValue, false, {CelValue::Type::kString}));
+
+  // static functions
+  addStaticFunctionToRegistry(registry, std::make_unique<UrlFunction>(StaticFuncNameUrl));
+}
+
+CustomCelVocabularyPtr ExtendedRequestCelVocabularyFactory::createCustomCelVocabulary(
+    const Protobuf::Message& config, ProtobufMessage::ValidationVisitor& validation_visitor) {
+  ExtendedRequestCelVocabularyConfig custom_cel_config =
+      MessageUtil::downcastAndValidate<const ExtendedRequestCelVocabularyConfig&>(
+          config, validation_visitor);
+  return std::make_unique<ExtendedRequestCelVocabulary>(
+      custom_cel_config.return_url_query_string_as_map());
+}
+
+REGISTER_FACTORY(ExtendedRequestCelVocabularyFactory, CustomCelVocabularyFactory);
```

Comment:
@ashankar1 this is in a `common` directory, but is registering a specific extension..